### PR TITLE
Matching L1 to PF Jets for HLT_VBF paths.

### PR DESCRIPTION
**Pull request by vukasinmilosevic**
Matching L1 to PF Jets. Used for HLT_VBF paths.
*Matches PF Jets to L1 jets from the seed
*Adds selection criteria to the leading/subleading jets as well as the maximum dijet mass
*Separates collections of PF jets into two categories

### CODE REVIEW

File: RecoTauTag/HLTProducers/interface/L1PFJetsMatching.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+ private:
+    
+  const edm::EDGetTokenT<reco::PFJetCollection> jetSrc;
+  const edm::EDGetTokenT<trigger::TriggerFilterObjectWithRefs> jetTrigger;
+  const double pt1_Min;
+  const double pt2_Min;
+  const double mjj_Min;
+
```
Comment (fwyzard) [inline RecoTauTag/HLTProducers/interface/L1PFJetsMatching.h]: please follow CMS conding rules: member variables should be called `jetSrc_` or `m_jetSrc`

File: RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+//
+// class decleration
+//
+using namespace reco   ;
+using namespace std    ;
+using namespace edm    ;
+using namespace trigger;
+
+std::pair<PFJetCollection,PFJetCollection> Categorise(PFJetCollection CaloL2jets, double pt1, double pt2, double Mjj)
```
Comment (fwyzard) [inline RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc]: I don't know if we have a rule for it, but I think `Categorise` should be lower case

File: RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+{
+
+int A=0;
```
Comment (fwyzard) [inline RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc]: local variables should be lower case; also please use a meaningful name

File: RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    std::pair<PFJetCollection,PFJetCollection> output;
+   unsigned int i1 = 0;
+   unsigned int i2 = 0;
+    double mjj = -9999;
```
Comment (fwyzard) [inline RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc]: start from 0.

File: RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    if (CaloL2jets.size()>1){
+    for (unsigned int i = 0; i < CaloL2jets.size()-1; i++)
+        for (unsigned int j = i+1; j < CaloL2jets.size(); j++)
+    {
+        const PFJet &  myJet1 = (CaloL2jets)[i];
+        const PFJet &  myJet2 = (CaloL2jets)[j];
+
```
Comment (fwyzard) [inline RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc]: please clean up the extra empty lines

File: RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        
+        
+        if ((myJet1.p4()+myJet2.p4()).M()>mjj){
+            
+         mjj =(myJet1.p4()+myJet2.p4()).M();
+            i1 = i;
+            i2 = j;
+        }
+    }
+        //std::cout<<"mjj:= "<<mjj<<"; i1,i2: = "<<i1<<" "<<i2<<std::endl;
```
Comment (fwyzard) [inline RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc]: please remove the commented-out code

File: RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc (hunk 3 above)
Comment (fwyzard) [inline RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc]: in fact I think `A` is actually unused ?

File: RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+  produces<PFJetCollection>("TwoJets");
+  produces<PFJetCollection>("ThreeJets");
+}
+L1PFJetsMatching::~L1PFJetsMatching(){ }
+
+void L1PFJetsMatching::produce(edm::StreamID iSId, edm::Event& iEvent, const edm::EventSetup& iES) const
+{
+    
+  unique_ptr<PFJetCollection> caloL2jets(new PFJetCollection);
+    std::pair<PFJetCollection,PFJetCollection> Output;
```
Comment (fwyzard) [inline RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc]: lowercase `output`

File: RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+       // }
+        break;
+      }
+    }
+  }  
+   
+    
+    
+    
+    Output= Categorise(*caloL2jets,pt1_Min,pt2_Min, mjj_Min);
```
Comment (fwyzard) [inline RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc]: you can move the declaration here:
```
auto output = categorise(...
```

File: RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+            Output.first.push_back(myJet1);
+            Output.first.push_back(myJet2);
+            
+	}
+        
+        if ((myJet1.p4().Pt() < pt1) && (myJet1.p4().Pt() > pt2) && (myJet2.p4().Pt() > pt2) && (mjj > Mjj))
+        {
+            
+            const PFJet &  myJetTest = (CaloL2jets)[0];
+         if (myJetTest.p4().Pt()>90){
```
Comment (rmanzoni) [inline RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc]: pt 90 GeV shouldn't be hardcoded, but configurable. I guess it should be pt1_Min, correct?

File: RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc (hunk 5 above)
Comment (rmanzoni) [inline RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc]: while we're at it, CaloL2jets is not a good name for a PFJetCollection 
(although I know why it ended up being named this way)

File: RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    iEvent.put(std::move(output1),"TwoJets");
+    iEvent.put(std::move(output2),"ThreeJets");
+
+}
+
+void L1PFJetsMatching::fillDescriptions(edm::ConfigurationDescriptions& descriptions) 
+{
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("L1JetTrigger", edm::InputTag("hltL1sDoubleIsoTau40er"                     ))->setComment("Name of trigger filter"    );
+  desc.add<edm::InputTag>("JetSrc"      , edm::InputTag("hltSelectedCaloJetsTrackPt1MediumIsolationReg"))->setComment("Input collection of PFJets");
```
Comment (rmanzoni) [inline RecoTauTag/HLTProducers/src/L1PFJetsMatching.cc]: put meaningful defaults for the input collections, for example the name of you VBF seed filter and the name of the PFJetCollection that you use in the matching

**Review by rmanzoni — COMMENTED** [review]
one hard coded value to make configurable and some minor style comments

**Comment by silviodonato** [comment]
Please test

**Comment by rmanzoni** [comment]
hey,
I just noticed this PR goes against CMSSW's 920_pathX branch.
If I recall correctly though, PRs should go against master branch.

@fwyzard what should we do?

**Comment by fwyzard** [comment]
On 26 May 2017 at 17:27, Riccardo Manzoni <notifications@github.com> wrote:

> I just noticed this PR goes against CMSSW's 920_pathX branch.
> If I recall correctly though, PRs should go against master branch.
>
> <https://github.com/fwyzard>
>
I think it should be for the "master" branch, since it will go along with
all the other HLT developments that will require a new 9.2.x release.

.A

-- 
https://www.youtube.com/embed/oshus9DxEG8

**Comment by rmanzoni** [comment]
ok, then I think @vukasinmilosevic can open another, identical PR pointing at the master branch.

**Comment by Martin-Grunewald** [comment]
please test

**Comment by davidlange6** [comment]
move development to master branch. Thanks

**Comment by fwyzard** [comment]
@vukasinmilosevic I think you can simply click `Edit` at the top of this page and change the target branch to `master`.

**Comment by Martin-Grunewald** [comment]
@vukasinmilosevic 
you still need to re-open this PR!

**Comment by vukasinmilosevic** [comment]
Started a new pull request, #18992 (couldn't reopen this one)

**Pull request closed without merging**